test/TokenPaymaster.test.js

```diff
@@ -0,0 +1,228 @@
+        ...sharedRelayRequestData
+      })
+      const ret = await retcode(paymaster.acceptRelayedCall(relayRequest, '0x', 1e6))
+      assert.equal(ret.code, 0, ret)
+    })
+  })
+
+  context('relayedCall', () => {
+    const paymasterDeposit = 1e18.toString()
+
+    before(async () => {
+      await hub.stake(relay, 7 * 24 * 3600, {
+        from: relayOwner,
+        value: 2e18
+      })
+      await hub.registerRelay(2e16.toString(), '10', 'url', { from: relay })
+      await hub.depositFor(paymaster.address, { value: paymasterDeposit })
+    })
+
+    it('pay with token to make a call', async () => {
+      const preTokens = await token.balanceOf(recipient.address)
+      const prePaymasterTokens = await token.balanceOf(paymaster.address)
+      // for simpler calculations: we don't take any fee, and gas price is '1', so actual charge
+      // should be exactly gas usage. token is 2:1 to eth, so we expect to pay exactly twice the "charge"
+      const relayRequest = new RelayRequest({
+        ...sharedRelayRequestData,
+        senderAddress: from,
+        senderNonce: (await hub.getNonce(from)).toString(),
+        gasPrice: '1',
+        pctRelayFee: '0',
+        baseRelayFee: '0'
+      })
+
+      const chainId = await web3.eth.net.getId()
+      const { signature } = await getEip712Signature({
+        web3,
+        chainId,
+        relayHub: hub.address,
+        relayRequest
+      })
+      const maxGas = 1e6
+      const arcGasLimit = 1e6
+
+      // not really required.
+      assert.deepEqual(await retcode(hub.canRelay(relayRequest, maxGas, arcGasLimit, signature, '0x', {
+        from: relay,
+        gasPrice: 1
+      })), {
+        code: 0,
+        msg: ''
+      })
+      const preBalance = await hub.balanceOf(paymaster.address)
+
+      const ret = await hub.relayCall(relayRequest, signature, '0x', {
+        from: relay,
+        gasPrice: 1,
+        gas: 1.516181e6
+      })
+      console.log('rh.deposit before', (await hub.balanceOf(paymaster.address)).toString())
+
+      // console.log(getLogs(ret))
+      const relayed = ret.logs.find(log => log.event === 'TransactionRelayed')
+      const events = await paymaster.getPastEvents()
+      console.log(getLogs(events))
+      const chargedEvent = events.find(e => e.event === 'TokensCharged')
+
+      assert.equal(relayed.args.status, 0)
+      const postTokens = await token.balanceOf(recipient.address)
+      const usedTokens = preTokens - postTokens
+
+      console.log('recipient tokens balance change (used tokens): ', usedTokens.toString())
+      console.log('reported charged tokens: ', chargedEvent.args.tokenActualCharge.toString())
+      const expectedTokenCharge = await uniswap.getTokenToEthOutputPrice(chargedEvent.args.ethActualCharge)
+      assert.closeTo(usedTokens, expectedTokenCharge.toNumber(), 1000)
+      const postBalance = await hub.balanceOf(paymaster.address)
+
+      assert.ok(postBalance >= preBalance,
```

Comment:
there is 'lte' I assume

Reply:
there is "isAtLeast", but it doesn't work well with bignumbers..